File: 2011/S113.cpp
#include <cstdio>
#include <cstring>
char a[1002][1002];
int up[1002][1002], down[1002][1002], left[1002][1002], right[1002][1002];
char s[4000001];
int main() {
  int n, m, c;
  scanf("%d %d %d", &n, &m, &c);
  for (int i = 1; i <= n; i++) {
    scanf("%s", a[i] + 1);
    a[i][0] = a[i][m + 1] = '#';
  }
  for (int j = 1; j <= m; j++) {
    a[0][j] = a[n + 1][j] = '#';
  }
  for (int i = 1; i <= n; i++) {
    for (int j = 1; j <= m; j++) {
      left[i][j] = a[i][j - 1] == '#' ? 0 : left[i][j - 1] + 1;
    }
    for (int j = m; j >= 1; j--) {
      right[i][j] = a[i][j + 1] == '#' ? 0 : right[i][j + 1] + 1;
    }
  }
  for (int j = 1; j <= m; j++) {
    for (int i = 1; i <= n; i++) {
      up[i][j] = a[i - 1][j] == '#' ? 0 : up[i - 1][j] + 1;
    }
    for (int i = n; i >= 1; i--) {
      down[i][j] = a[i + 1][j] == '#' ? 0 : down[i + 1][j] + 1;
    }
  }
  long long sum = 0;
  int x, y;
  scanf("%d %d %s", &x, &y, s);
  for (int i = 0; i < c; i++) {
    if (s[i] == 'N') {
      sum += up[x][y];
      x -= up[x][y];
    } else if (s[i] == 'E') {
      sum += right[x][y];
      y += right[x][y];
    } else if (s[i] == 'S') {
      sum += down[x][y];
      x += down[x][y];
    } else {
      sum += left[x][y];
      y -= left[x][y];
    }
  }
  printf("%lld\n", sum);
  return 0;
}

File: 2011/S113.pas
var
  a: array[0..1001, 0..1001] of char;
  up, down, left, right: array[0..1001, 0..1001] of longint;
  s: ansistring;
  n, m, c, i, j, x, y: longint;
  sum: int64;
begin
  readln(n, m, c);
  for i := 1 to n do
  begin
    for j := 1 to m do
      read(a[i][j]);
    a[i][0] := '#';
    a[i][m + 1] := '#';
    readln
  end;
  for j := 1 to m do
  begin
    a[0][j] := '#';
    a[n + 1][j] := '#';
  end;
  for i := 1 to n do
  begin
    left[i][0] := 0;
    for j := 1 to m do
      if (a[i][j - 1] = '#') then
        left[i][j] := 0
      else
        left[i][j] := left[i][j - 1] + 1;
    right[i][m + 1] := 0;
    for j := m downto 1 do
      if (a[i][j + 1] = '#') then
        right[i][j] := 0
      else
        right[i][j] := right[i][j + 1] + 1
  end;
  for j := 1 to m do
  begin
    up[0][j] := 0;
    for i := 1 to n do
      if (a[i - 1][j] = '#') then
        up[i][j] := 0
      else
        up[i][j] := up[i - 1][j] + 1;
    down[n + 1][j] := 0;
    for i := n downto 1 do
      if (a[i + 1][j] = '#') then
        down[i][j] := 0
      else
        down[i][j] := down[i + 1][j] + 1;
  end;
  sum := 0;
  readln(x, y);
  read(s);
  for i := 1 to c do
  begin
    if (s[i] = 'N') then
    begin
      sum := sum + up[x][y];
      x := x - up[x][y]
    end
    else if (s[i] = 'E') then
    begin
      sum := sum + right[x][y];
      y := y + right[x][y]
    end
    else if (s[i] = 'S') then
    begin
      sum := sum + down[x][y];
      x := x + down[x][y]
    end
    else
    begin
      sum := sum + left[x][y];
      y := y - left[x][y]
    end
  end;
  writeln(sum)
end.
